core/src/main/scala/cats/Monad.scala

```diff
@@ -14,4 +14,36 @@ import simulacrum.typeclass
 @typeclass trait Monad[F[_]] extends FlatMap[F] with Applicative[F] {
   override def map[A, B](fa: F[A])(f: A => B): F[B] =
     flatMap(fa)(a => pure(f(a)))
+
+  /**
+   * Execute an action repeatedly as long as the given `Boolean` expression
+   * returns `true`. The condition is evalated before the loop body.
+   * Collects the results into an arbitrary `MonadCombine` value, such as a `List`.
+  */
+  def whileM[G[_], A](p: F[Boolean])(body: => F[A])(implicit G: MonadCombine[G]): F[G[A]] = {
```

Comment:
@johnynek I think the intention here is to avoid creating the body `F[A]` if `p` is never true. Thanks to you, I noticed that I missed the memoization of body. I just implemented using a `Later`.
@edmundnoble internally we don't use by-name params, but we do have plenty of public APIs using it for the sake of ease to use when we don't really want to give user the control over evaluation. 
In this particular case, having user passing in a by-name and then memoize it might be misleading. I am open to change to take in a `Later[A]` instead. Or maybe it's okay to give user the control and just take in an `Eval`. What do you guys think?